src/WeakRefStrings.jl

```diff
@@ -141,4 +185,184 @@ function Base.vcat(a::WeakRefStringArray{T, 1}, b::WeakRefStringArray{T, 1}) whe
+        end
+    end
+    s
+end
+Base.convert(::Type{StringArray}, arr::AbstractArray{T}) where {T<:STR} = StringArray{T}(arr)
+Base.convert(::Type{StringArray{T, N} where T}, arr::AbstractArray{S}) where {S<:STR, N} = StringVector{S}(arr)
+(::Type{StringVector{T}})() where {T} = StringVector{T}(Vector{UInt8}(0), UInt64[], UInt32[])
+(::Type{StringVector})() = StringVector{String}()
+
+_isassigned(arr, i...) = isassigned(arr, i...)
+_isassigned(arr, i::CartesianIndex) = isassigned(arr, i.I...)
+@inline Base.@propagate_inbounds function Base.getindex(a::StringArray{T}, i::Integer...) where T
+    offset = a.offsets[i...]
+    if offset == UNDEF_OFFSET
+        throw(UndefRefError())
+    end
+
+    if Missing <: T && offset === MISSING_OFFSET
+        return missing
+    end
+
+    convert(T, WeakRefString(pointer(a.buffer) + offset, a.lengths[i...]))
+end
+
+function Base.similar(a::StringArray, T::Type{<:STR}, dims::Tuple{Vararg{Int64, N}}) where N
+    StringArray{T, N}(dims)
+end
+
+function Base.empty!(a::StringVector)
+    empty!(a.buffer)
+    empty!(a.offsets)
+    empty!(a.lengths)
+end
+
+Base.copy(a::StringArray{T, N}) where {T,N} = StringArray{T, N}(copy(a.buffer), copy(a.offsets), copy(a.lengths))
+
+@inline function Base.setindex!(arr::StringArray, val::WeakRefString, idx::Integer...)
+    p = pointer(arr.buffer)
+    if val.ptr <= p + sizeof(arr.buffer)-1 && val.ptr >= p
+        # this WeakRefString points to data entirely within arr's buffer
+        # don't add anything to the buffer in this case.
+        # this optimization helps `permute!`
+        arr.offsets[idx...] = val.ptr - p
+        arr.lengths[idx...] = val.len
+    else
+        _setindex!(arr, val, idx...)
+    end
+end
+
+@inline function Base.setindex!(arr::StringArray, val::STR, idx::Integer...)
+    _setindex!(arr, val, idx...)
+end
+
+function _setindex!(arr::StringArray, val::AbstractString, idx...)
+    buffer = arr.buffer
+    l = length(arr.buffer)
+    resize!(buffer, l + sizeof(val))
+    unsafe_copy!(pointer(buffer, l+1), pointer(val,1), sizeof(val))
+    arr.lengths[idx...] = sizeof(val)
+    arr.offsets[idx...] = l
+    val
+end
+
+function _setindex!(arr::StringArray, val::Missing, idx...)
+    arr.lengths[idx...] = 0
+    arr.offsets[idx...] = MISSING_OFFSET
+end
+
+function Base.resize!(arr::StringVector, len)
+    l = length(arr)
+    resize!(arr.offsets, len)
+    resize!(arr.lengths, len)
+    if l < len
+        arr.offsets[l+1:len] = UNDEF_OFFSET # undef
+        arr.lengths[l+1:len] = 0
+    end
+    arr
+end
+
+function Base.push!(arr::StringVector, val::AbstractString)
+    l = length(arr.buffer)
+    resize!(arr.buffer, l + sizeof(val))
+    unsafe_copy!(pointer(arr.buffer, l + 1), pointer(val,1), sizeof(val))
+    push!(arr.offsets, l)
+    push!(arr.lengths, sizeof(val))
+    arr
+end
+
+function Base.push!(arr::StringVector, val::Missing)
+    push!(arr.offsets, MISSING_OFFSET)
+    push!(arr.lengths, 0)
+end
+
```

Comment:
Can you add a method for `deleteat!` on StringVector? It just came up as useful in #28 for WeakRefStringArray too.